src/utils/generalFunctions.js

```diff
@@ -67,14 +68,76 @@ export function aboutPopup(call, classes = {}) {
   });
 }
 
+function breakToolId(tabId) {
+  const index = tabId.search(/\d/);
+  const alpha = tabId.substring(0, index - 1);
+
+  if (index < 0)
+    return [alpha];
+
+  const number = new Number(tabId.substring(index));
+
+  return [alpha, number];
+}
+
+function computeIds(tabsMap) {
+  const res = {};
+
+  for (const [tabId] of tabsMap) {
+    const [alpha, number] = breakToolId(tabId);
+    const index = tabId.search(/\d/);
+    if (index < 0)
+      continue;
+    const specific = res[alpha] ?? { last: 0, busy: {}, free: [] };
+    specific.busy[number] = 1;
+    if (number > specific.last)
+      specific.last = number;
+    res[alpha] = specific;
+  }
+
+  for (const toolName in res) {
+    const specific = res[toolName];
+    for (let j = 0; j < specific.last; j++)
+      if (!specific.busy[j])
+        specific.free.push(j);
+    delete specific.busy;
+    specific.last++;
+  }
+
+  return res;
+}
+
+const workspace = new Workspace();
+const tabsMap = workspace.getTabs();
+const toolIds = computeIds(tabsMap);
+
+export function freeToolId(tabId) {
+  const [alpha, number] = breakToolId(tabId);
+
+  if (!number)
+    return;
+
+  const toolIdData = toolIds[alpha];
+  toolIdData.free.unshift(number);
+}
+
```

Comment:
Actually coming to review this again, I'm trying to understand this code, and looks so much more, less performant and more complex than what I did [here](https://github.com/MOV-AI/frontend-npm-lib-ide/pull/334/files) in the `src/utils/Utils.js` to achieve the same thing. Can you confirm @EdwardAngeles @quirinpa ?

Reply:
That only runs when the page loads.
It is to compute the things instead of using localStorage.
After the initial time it runs, it's pretty fast.